smarts/core/waymo_map.py

```diff
@@ -705,14 +811,27 @@ def parse_source_to_scenario(source: str):
         """Read the dataset file and get the specified scenario"""
         dataset_path = source.split("#")[0]
         scenario_id = source.split("#")[1]
-        dataset_records = read_tfrecord_file(dataset_path)
-        for record in dataset_records:
+
+        # Reset cache if this is a new TFRecord file
+        if not WaymoMap._tfrecord_path or WaymoMap._tfrecord_path != dataset_path:
+            WaymoMap._tfrecord_path = dataset_path
+            WaymoMap._tfrecord_generator = read_tfrecord_file(dataset_path)
+            WaymoMap._scenario_cache = dict()
+
+        if scenario_id in WaymoMap._scenario_cache:
+            return WaymoMap._scenario_cache[scenario_id]
```

Comment:
It's minor here, I know, but this pattern avoids double lookups in dicts:
```suggestion
        parsed_scenario = WaymoMap._scenario_cache.get(scenario_id)
        if parsed_scenario:
            return parsed_scenario
```